[1.12] Added docs for policy exceptions support in the CLI. (#1137)
kyverno/website: content/en/docs/Kyverno CLI/_index.md

kyverno apply /path/to/policy1.yaml /path/to/folderFullOfPolicies --resource /path/to/resource1.yaml --resource /path/to/resource2.yaml --cluster
```

> Apply a policy to a resource with exception: 

Reviewer: ```suggestion
Apply a policy to a resource with a policy exception: 
```

  warn: 0
```

>  Example: 

Reviewer: Yes, this should all be in a new section.

  warn: 0
```

#### Applying Policy Exceptions 

> Apply a policy manifest to a resource manifest with a policy exception, using `--exceptions` flag to pass the Policy Exception manifest.

Reviewer: ```suggestion
[Policy Exceptions](/docs/writing-policies/exceptions/) can be applied alongside policies by using the `--exceptions` flag to pass the Policy Exception manifest.
```

```sh
kyverno apply /path/to/policy.yaml --resource /path/to/resource.yaml --exception /path/to/exception.yaml
```

Example:

>  Applying Policy on a resource with a policy exception 

Reviewer: ```suggestion
Applying a policy to a resource with a policy exception 
```

>  Policy manifest (`policy.yaml`):

Reviewer: ```suggestion
Policy manifest (`policy.yaml`):
```

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: max-containers
spec:
  validationFailureAction: Enforce
  background: false
  rules:
  - name: max-two-containers
    match:
      any:
      - resources:
          kinds:
          - Pod
    validate:
      message: "A maximum of 2 containers are allowed inside a Pod."
      deny:
        conditions:
          any:
          - key: "{{request.object.spec.containers[] | length(@)}}"
            operator: GreaterThan

>             value: "2"

Reviewer: ```suggestion
            value: 2
```

    - max-two-containers
    - autogen-max-two-containers
  match:
    any:
    - resources:
        kinds:
        - Pod
        - Deployment
  conditions:
    any:
    - key: "{{ request.object.metadata.labels.color || '' }}"
      operator: Equals
      value: blue
```

Resource manifest (`resource.yaml`):

A Deployment matching the characteristics defined in the PolicyException, shown below, will be allowed creation even though it technically violates the rule’s definition.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: three-containers-deployment
  labels:
    app: my-app
    color: blue
spec:
  replicas: 3
  selector:
    matchLabels:
      app: my-app
  template:
    metadata:
      labels:
        app: my-app
        color: blue
    spec:
      containers:
        - name: nginx-container
          image: nginx:latest
          ports:
            - containerPort: 80
        - name: redis-container
          image: redis:latest
          ports:
            - containerPort: 6379
        - name: busybox-container
          image: busybox:latest
          command: ["/bin/sh", "-c", "while true; do echo 'Hello from BusyBox'; sleep 10; done"]    
```

> Apply the above policy on resource with exception 

Reviewer: ```suggestion
Apply the above policy to the resource with the exception 
```

resources:
  - <path/to/resource.yaml>
  - <path/to/resource.yaml>

> exceptions: # optional file for specifying exceptions. see below for example.

Reviewer: ```suggestion
exceptions: # optional files for specifying exceptions. See below for an example.
```

1. The `policies` element which lists one or more policies to be applied.
2. The `resources` element which lists one or more resources to which the policies are applied.

> 3. The `exceptions` element which lists one or more policy exceptions to the policies and can't be used along with ValidatingAdmissionPolicy. Optional.

Reviewer: ```suggestion
3. The `exceptions` element which lists one or more policy exceptions. Cannot be used with ValidatingAdmissionPolicy. Optional.
```

Test Summary: 1 tests passed and 0 tests failed
```

> In the following policy test, a `validate` policy rule ensures that Pods isn't allowed to access the host namespaces except for those named `important-tool*` in the `delta` namespace. This is achieved by creating a Policy Exception and defining its file path in the `exceptions` field in the Test manifest. It is expected that resources that violate the policy rule but match policy exceptions will be skipped. Otherwise, they will fail.

Reviewer: ```suggestion
In the following policy test, a `validate` rule ensures that Pods aren't allowed to access host namespaces. A Policy Exception is used to exempt Pods and Deployments beginning with the name `important-tool` in the `delta` namespace from this rule. The `exceptions` field is used in the Test manifest to declare a Policy Exception manifest. It is expected that resources that violate the rule but match policy exceptions will be skipped. Otherwise, they will fail.
```

Policy manifest (`disallow-host-namespaces.yaml`):

```yaml
apiVersion: kyverno.io/v2beta1
kind: ClusterPolicy
metadata:
  name: disallow-host-namespaces
spec:
  validationFailureAction: Enforce
  background: false
  rules:
    - name: host-namespaces
      match:
        any:
        - resources:
            kinds:
              - Pod
      validate:
        message: >-
          Sharing the host namespaces is disallowed. The fields spec.hostNetwork,
          spec.hostIPC, and spec.hostPID must be unset or set to `false`.          
        pattern:
          spec:
            =(hostPID): "false"
            =(hostIPC): "false"
            =(hostNetwork): "false"
```

Policy Exception manifest (`delta-exception.yaml`):

```yaml
apiVersion: kyverno.io/v2beta1
kind: PolicyException
metadata:
  name: delta-exception
  namespace: delta
spec:
  exceptions:
  - policyName: disallow-host-namespaces
    ruleNames:
    - host-namespaces
    - autogen-host-namespaces
  match:
    any:
    - resources:
        kinds:
        - Pod
        - Deployment
        namespaces:
        - delta
        names:
        - important-tool*
```

Resource manifest (`resource.yaml`):

> Both deployments violate the policy rule but one of which matches an exception. Thus it will be successfully created.

Reviewer: ```suggestion
Both Deployments violate the policy but only one matches an exception. The Deployment without an exception will fail while the one with an exception will be skipped.
```
